File: Algorithm/060. Permutation Sequence.py
# -*- coding: utf-8 -*-
# @Author: WuLC
# @Date:   2016-05-08 09:31:46
# @Last modified by:   WuLC
# @Last Modified time: 2016-05-08 15:23:48
# @Email: liangchaowu5@gmail.com


# 利用nextPermutation，TLE >_<
class Solution(object):
    def getPermutation(self, n, k):
        """
        :type n: int
        :type k: int
        :rtype: str
        """
        nums = [i for i in xrange(1,n+1)]
        for i in xrange(k-1):
            self.nextPermutation(nums)
        return reduce(lambda x,y:str(x)+str(y),nums)

    def nextPermutation(self, nums):
        sortList = sorted(nums)
        if nums == sortList[::-1]:
            nums.sort()
            return
        
        i = len(nums)-1
        flag = 0
        while i > 0:
            if nums[i] > nums[i-1]:
                # 往后找大于nums[i-1]且小于nums[i]的数与nums[i-1]交换
                j = len(nums)-1
                while j>i:
                    if nums[i-1]<nums[j]<nums[i]:
                        nums[j],nums[i-1] = nums[i-1],nums[j]
                        flag = 1
                        break
                    j-=1
                # 找不到这样的数
                if flag == 0:
                    nums[i],nums[i-1] = nums[i-1],nums[i]
                # 交换后对i后面的元素进行排序
                nums[i:] = sorted(nums[i:])
                return                 
            else:
                i-=1


# 参考https://leetcode.com/discuss/47086/share-my-python-solution-with-detailed-explanation，AC 
class Solution(object):
    def getPermutation(self, n, k):
        """
        :type n: int
        :type k: int
        :rtype: str
        """
        nums = range(1,n+1)
        k -= 1  # 不减1会导致k为最后一个数时越界
        res = ''
        while n > 0:
            n -= 1
            index, k = divmod(k,math.factorial(n)) # 0! = 1，0%n = 0
            res += str(nums.pop(index))
        return res
        

File: Algorithm/Java/060. Permutation Sequence.java
/**
* Author: WuLC
* Date:   2017-01-02 21:17:56
* Last modified by:   WuLC
* Last Modified time: 2017-01-02 21:26:03
* Email: liangchaowu5@gmail.com
*/


// locate each number bit by bit 
public class Solution 
{
    private String result;
    private int used;
    
    public String getPermutation(int n, int k) 
    {
        result = "";
        used = 0;
        dfs(n, k-1, 1);
        return result;
    }
    
    
    private void dfs(int n, int k, int m)
    {
        if (m>n) return;
        int tmp = factorial(n-m);
        int count = k/tmp, idx = 1, tmpCount = 0;
        k -= count*tmp;
        count += 1;
        for(; idx <= n; idx++)
        {
            if (((used>>idx) & 1) == 0) tmpCount++;
            if (tmpCount == count) break;
        }
        result += String.valueOf(idx);
        used ^= (1<<idx);
        dfs(n, k, m+1);
    }
    
    
    private int factorial(int n)
    {
        int fac = 1;
        for(int i=2; i<=n; i++) 
            fac *= i;
        return fac;
    }
}
